File: permutation_constraint.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<pair<int, int>> constraint;
bool chosen[21];

void permute(int n, vector<int> &p)
{
    if (p.size() == n)
    {
        for (int i : p)
        {
            cout << i << ' ';
        }
        cout << '\n';
    }
    else
    {
        for (int i = 0; i < n; ++i)
        {
            if (chosen[i])
                continue;
            bool con = false;
            for (auto &c : constraint)
            {
                if (i == c.second && !chosen[c.first])
                {
                    con = true;
                    break;
                }
            }
            if (con)
                continue;
            chosen[i] = true;
            p.push_back(i);
            permute(n, p);
            chosen[i] = false;
            p.pop_back();
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    int n, k;
    cin >> n >> k;
    while (k--)
    {
        int a, b;
        cin >> a >> b;
        constraint.emplace_back(a, b);
    }
    for (int i = 0; i < n; ++i)
        chosen[i] = false;
    std::vector<int> result;
    permute(n, result);
}

File: permutation_constraint.py
chosen = {}
constraint = []


def permute(k, p: list = []):
    if len(p) == k:
        print("".join(p))
        return
    for i in range(k):
        if chosen[i]:
            continue
        con = False
        for c in constraint:
            if i == c[1] and not chosen[c[0]]:
                con = True
                break
        if (con):continue
        chosen[i] = True
        p.append(str(i))
        permute(k, p)
        chosen[i] = False
        p.pop()


n, k = [int(e) for e in input().split()]
for _ in range(k):
    constraint.append([int(e) for e in input().split()])
for i in range(n):
    chosen[i] = False
permute(n)
